pkg/cmd/run/watch/watch.go

```diff
@@ -0,0 +1,228 @@
+	"github.com/cli/cli/pkg/cmd/run/shared"
+	"github.com/cli/cli/pkg/cmdutil"
+	"github.com/cli/cli/pkg/iostreams"
+	"github.com/cli/cli/utils"
+	"github.com/spf13/cobra"
+)
+
+const defaultInterval int = 3
+
+type WatchOptions struct {
+	IO         *iostreams.IOStreams
+	HttpClient func() (*http.Client, error)
+	BaseRepo   func() (ghrepo.Interface, error)
+
+	RunID      string
+	Interval   int
+	ExitStatus bool
+
+	Prompt bool
+
+	Now func() time.Time
+}
+
+func NewCmdWatch(f *cmdutil.Factory, runF func(*WatchOptions) error) *cobra.Command {
+	opts := &WatchOptions{
+		IO:         f.IOStreams,
+		HttpClient: f.HttpClient,
+		Now:        time.Now,
+	}
+
+	cmd := &cobra.Command{
+		Use:   "watch <run-selector>",
+		Short: "Runs until a run completes, showing its progress",
+		Annotations: map[string]string{
+			"IsActions": "true",
+		},
+		RunE: func(cmd *cobra.Command, args []string) error {
+			// support `-R, --repo` override
+			opts.BaseRepo = f.BaseRepo
+
+			if len(args) > 0 {
+				opts.RunID = args[0]
+			} else if !opts.IO.CanPrompt() {
+				return &cmdutil.FlagError{Err: errors.New("run ID required when not running interactively")}
+			} else {
+				opts.Prompt = true
+			}
+
+			if runF != nil {
+				return runF(opts)
+			}
+
+			return watchRun(opts)
+		},
+	}
+	cmd.Flags().BoolVar(&opts.ExitStatus, "exit-status", false, "Exit with non-zero status if run fails")
+	cmd.Flags().IntVarP(&opts.Interval, "interval", "i", defaultInterval, "Refresh interval in seconds")
+
+	return cmd
+}
+
+func watchRun(opts *WatchOptions) error {
+	c, err := opts.HttpClient()
+	if err != nil {
+		return fmt.Errorf("failed to create http client: %w", err)
+	}
+	client := api.NewClientFromHTTP(c)
+
+	repo, err := opts.BaseRepo()
+	if err != nil {
+		return fmt.Errorf("failed to determine base repo: %w", err)
+	}
+
+	runID := opts.RunID
+	var run *shared.Run
+
+	if opts.Prompt {
+		cs := opts.IO.ColorScheme()
+		runs, err := shared.GetRunsWithFilter(client, repo, 10, func(run shared.Run) bool {
+			return run.Status != shared.Completed
+		})
+		if err != nil {
+			return fmt.Errorf("failed to get runs: %w", err)
+		}
+		if len(runs) == 0 {
+			return fmt.Errorf("found no in progress runs to watch")
+		}
+		runID, err = shared.PromptForRun(cs, runs)
+		if err != nil {
+			return err
+		}
+		// TODO silly stopgap until dust settles and PromptForRun can just return a run
+		for _, r := range runs {
+			if fmt.Sprintf("%d", r.ID) == runID {
+				run = &r
+				break
+			}
+		}
+	} else {
+		run, err = shared.GetRun(client, repo, runID)
+		if err != nil {
+			return fmt.Errorf("failed to get run: %w", err)
+		}
+	}
+
+	if run.Status == shared.Completed {
+		return nil
+	}
+
+	prNumber := ""
+	number, err := shared.PullRequestForRun(client, repo, *run)
+	if err == nil {
+		prNumber = fmt.Sprintf(" #%d", number)
+	}
+
+	if runtime.GOOS == "windows" {
+		opts.IO.EnableVirtualTerminalProcessing()
+	}
+	// clear entire screen
+	fmt.Fprintf(opts.IO.Out, "\x1b[2J")
```

Comment:
This seems like it could be useful to put into IOStreams as a function like `ClearScreen`. I don't think we do this anywhere else, should we leave it up to the user to clear the screen before running the command? 

Side question. What happens to the scroll back buffer in the terminal when we clear the screen like this?

Reply:
Until we have a need for it in other commands I don't feel any urgency around putting it in IOStreams. I do similar stuff in some of my easter egg commands but I'd rather just rewrite those to use `tview` instead of relying on code in iostreams.

Clearing the screen is key because later on we move the cursor up to the 0,0 point of the screen. This is critical for "animated" content since we can't just keep clearing. All clearing does is print a bunch of newlines (which answers your question about scrollback). To "animate" with repeated clears makes scrollback just awful.